src/ophyd_async/core/_detector.py

```diff
@@ -312,39 +332,50 @@ async def prepare(self, value: TriggerInfo) -> None:
 
     @AsyncStatus.wrap
     async def kickoff(self):
-        assert self._trigger_info, "Prepare must be called before kickoff!"
-        if self._iterations_completed >= self._trigger_info.iteration:
-            raise Exception(
+        assert (
+            self._trigger_info and self._number_of_triggers_iter
+        ), "Prepare must be called before kickoff!"
+        if self._frames_completed >= self._trigger_info.total_number_of_triggers:
+            raise RuntimeError(
                 f"Kickoff called more than the configured number of "
-                f"{self._trigger_info.iteration} iteration(s)!"
+                f"{self._trigger_info.total_number_of_triggers} iteration(s)!"
             )
-        self._iterations_completed += 1
+        self._frames_to_complete = next(self._number_of_triggers_iter)
+        self._frames_completed += self._frames_to_complete
```

Comment:
I've changed my mind on this https://github.com/bluesky/ophyd-async/pull/581#discussion_r1777342185